[bugfix] Sync the browser connections when changed from the select dialogs

elpaso opened the pull request:
For now, the sync works only in one direction: from the dialogs
to the browser instances

I'm not sure about the best approach to enable the sync from the browser to the dialogs, suggestions welcome.

@wonder-sk , can you please have a look?

## Update 17/07/2017

The PR now also handles the synchronization from the browser dialogs to the individual provider dialogs and to the other browsers.

## TODO

- [x] rename base class to `QgsAbstractDataSourceWidget`
- [x] make the ArcGis class private and remove from API
- [x] **This cannot be done because the enum is used in core and we don't want to inclue GUI from core, do we ?** ~~move WidgetMode enum could from QgsProviderRegistry to `QgsAbstractDataSourceWidget`~~
- [x] comment connections changed signal
- [x] move arcgis source select to provider folder
- [x] Move all signals into the base class

elpaso committed:
[bugfix] Sync the bowser connections when changed from the select dialogs

For now, the sync works only in one direction: from the dialogs
to the browser instances

wonder-sk commented on src/gui/qgsbrowserdockwidget.h:
```diff
@@ -102,6 +102,8 @@ class GUI_EXPORT QgsBrowserDockWidget : public QgsDockWidget, private Ui::QgsBro
     //! Emitted when drop uri list needs to be handled
     void handleDropUriList( const QgsMimeDataUtils::UriList & );
+    //! Connections changed in the browser
+    void connectionsChanged( );
```
Is this signal ever emitted from somewhere? I only see other signals connected to it...

elpaso replied:
@wonder-sk thanks for reviewing, I thinks this is a leftover from when I tried to make this work the other way round: refresh the dialogs when the browser panel changes (and hence it can be removed). 

As you noticed (and I mentioned in the PR introduction) this implementation only works one way: changes in the dialogs are propagated to the browser(s), and it does also work from one browser to another. 

I tried to find a good way to propagate the changes from the browsers to the provider's dialogs without success. 

I think we should merge this PR (after the changes you requested) and target the synchronization of the dialogs from the browser in a future work.

elpaso replied:
@wonder-sk I've found some time for an initial base class implementation and a test WMS implementation of the synchronization connections. It looks likes it's working as expected and synchronization works fine in both directions.

Would you please confirm that this is a good solution (for now WMS only!) before I spend more time refactoring the other source select widgets and data items?

I'm not convinced about the naming of the new class, I would have called it `QgsSourceSelectDialog` but that's already taken by the class that @manisandro  wrote for the ARC* providers. 

@manisandro if that class was not intended to be the real base class for all source select dialogs (it appears to be too specialized for online services), would you mind if I rename it to something more in line with its functionality as a base class for ARC* online services only? (suggestions welcome)

manisandro replied:
@elpaso The code actually comes from the WFS provider (as written at the time) and I had AFS, AMS and WFS using this base class, so not strictly for the arcgis providers. But indeed now the arcgis providers are the only ones still using it, so it might as well be moved to the arcgis provider folder and called QgsArcGisServiceSourceSelect or something similar, if it can't be used as a common base class for other online service dialogs anymore.

wonder-sk replied:
hi @elpaso - in general the new base class looks good to me - I have few suggestions though:
- can we switch from QDialog to QWidget as the base? this seems more appropriate when they are not used as standalone dialog anymore
- ```Qt::WindowFlags fl``` in the constructor is probably not necessary anymore (not used as standalone dialogs)
- ```WidgetMode``` enum could be probably moved from QgsProviderRegistry to this new class - maybe a more appropriate place?
- thinking about naming of the new class, how about something like ```QgsDataSourceWidget``` or ```QgsAbstractDataSourceWidget``` ? The name ```QgsSourceSelect``` does not say much...
- it would be nice to add a comment to ```connectionsChanged``` signal saying that it is used to refresh browser items

One more comment regarding the arcgis dialog - I would strongly prefer not to have it in public API - so I would suggest to either mark the class as private - or even better as Sandro suggested, to move it to arcgis provider where it belongs...

elpaso replied:
@wonder-sk thanks for the comments, I'll try to follow your suggestions, indeed I plan to move more common functionalities stuff into the base class as long as I move to refactor the other provider dialogs.

I'm not sure about dropping completely the possibility to use the dialogs as standalone, don't you think that they still could be useful in some particular situations where the type of layer is know (or constrained) in advance?

I'm blocked on the arcgis stuff though: I've already wasted quite some trying to move  the arcgis dialog to the provider as suggested by @manisandro but the canvas dependency stepped on my toes again: that dependency implies that the class must be available and included in the data source manager dialog code, that is in now in GUI, and I couldn't find the right way (well, *a* way) to include a provider's class (the moved arcgis dialog) from GUI. Maybe @manisandro (sorry Sandro to bother you again about this canvas dependency) knows a way to remove the dependency?

As a possible workaround, for example I was thinking of an additional (empty by default) virtual method in the base class, the sets the extent and CRS, that method would be called from the app (or whatever) to set the extent and CRS . This would also remove the canvas argument from the data source manager dialog, that is really only used for the arcgis provider.

wonder-sk commented on src/app/qgsdatasourcemanagerdialog.h:
```diff
@@ -79,6 +81,10 @@ class QgsDataSourceManagerDialog : public QgsOptionsDialogBase
+    //! Connections changed
+    void connectionsChanged( );
+    // internal signal
+    void dlg_refresh( );
```
Ideally this should not be handled by a signal, but rather through a common base class for dialogs - adding more signals like this will make the code difficult to read / understand. If starting a base class for the individual provider dialogs is too much of the work, I would suggest to at least rename the signal (e.g. providerDialogsRefreshRequested) and document it...

wonder-sk reviewed (commented):
Looks good in general, but see my inline comments... I think when connections are changed within browser widget they will not be picked up by data source manager dialogs

elpaso committed:
[bugfix] Sync the dialogs connections when changed from the browser

For now it's only for WMS but you get the idea.

There is a new abstract base class for the source select
dialogs, that will grow with common behavior for all
the select dialogs.

Signals are forwarded from the (root) data items to the
app and then delivered to the various browser instances
and to the unified layer dialog.

A change in one of the browser items should trigger a
refresh in all the other browsers and dialogs.

elpaso committed:
Added python bindings for the base source select class

elpaso committed:
SIP include

elpaso committed:
Added since 3.0

elpaso committed:
Renamed QgsSourceSelectDialog to QgsArcGisServiceSourceSelect

elpaso committed:
Added since 3.0 to the renamed class

elpaso committed:
Refactored all providers to use the new base class

Also use refreshConnections from the data items when
a refresh is required, this function also emits
the signal to update the other GUI elements.

elpaso committed:
Renamed base class to QgsAbstractDataSourceWidget

elpaso committed:
Complete documentation for connectionsChanged signals

elpaso committed:
Homogenize signals from different source select dialogs, move to base class

elpaso committed:
Moved arcgis dialogs to the provider directory

elpaso committed:
Store canvas into the base class and set extent/crs from the arcgis classes

This modification was necessary because the current implementation
of the source select dialogs within the unified add layer dialog
create the provider dialogs the first time and do not destroy
them, this means that the canvas extent and CRS can change from
a dialog invocation to the next and the extent and CRS need to
be updated at layer creation time.

nyalldawson commented on src/gui/qgsabstractdatasourcewidget.h:
```diff
@@ -0,0 +1,100 @@
+    void connectionsChanged();
+
+    //! Emitted when a DB layer has been selected for addition
+    void addDatabaseLayers( const QStringList &paths, const QString &providerKey );
```
I'm curious what the need is for this signal (vs addVectorLayer). What's the rationale here?

elpaso replied:
@nyalldawson  I've only ported that signal into the base class. I've already removed `addWfsLayer` signal that was really a clone of the vector signal, and maybe some other variations of the same.
I believe that during the years (because there was no base class/interface) every provider author just added his own signal semantic, now I'd like to merge all signals into a single one, but one step at a time.

btw, looking at the code, I guess the only purpose of that signal is that it is handled differently by the app: `void QgisApp::addDatabaseLayers( QStringList const &layerPathList, QString const &providerKey )`

nyalldawson commented:
@elpaso this is looking very nice!

elpaso committed:
Rebase and partially revert 9bae83275368

All signals are now in the base class, even if only
a subset of available providers actually emits them.

This way we can handle all source select dialogs
the same way, regardless if they are vector, DB
or raster (or others).

elpaso committed:
Remove unused includes and variables

elpaso commented:
@nyalldawson thanks :) I'm merging as it is, next step will be:
- really turn all dialogs into widgets
- get rid of unnecessary window flag
- create a raster source select dialog instead of forwarding to the app
- group all add* signals into a single `layersSelected` (this would make the data source manager dialog easily re-usable and fully decoupled from the app <- I still need to fully investigate how to do this one)
